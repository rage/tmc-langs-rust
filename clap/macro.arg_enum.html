<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Convenience macro to generate more complete enums with variants to be used as a type when parsing arguments. This enum also provides a `variants()` function which can be used to retrieve a `Vec&lt;&amp;&#x27;static str&gt;` of the variant names, as well as implementing `FromStr` and `Display` automatically."><title>arg_enum in clap - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-c5d6553a23f1e5a6.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="clap" data-themes="" data-resource-suffix="" data-rustdoc-version="1.81.0 (eeb90cda1 2024-09-04)" data-channel="1.81.0" data-search-js="search-d234aafac6c221dd.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-118b08c4c78b968e.js"></script><script defer src="sidebar-items.js"></script><script defer src="../static.files/main-d2fab2bf619172d3.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-df360f571f6edeae.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc macro"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../clap/index.html">clap</a><span class="version">2.34.0</span></h2></div><div class="sidebar-elems"></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Macro <a href="index.html">clap</a>::<wbr><a class="macro" href="#">arg_enum</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../src/clap/macros.rs.html#317-418">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><pre class="rust item-decl"><span class="macro">macro_rules!</span> arg_enum {
    (@as_item $(<span class="macro-nonterminal">$i</span>:item)<span class="kw-2">*</span>) =&gt; { ... };
    (@impls ( $(<span class="macro-nonterminal">$tts</span>:tt)* ) -&gt; (<span class="macro-nonterminal">$e</span>:ident, $(<span class="macro-nonterminal">$v</span>:ident),+)) =&gt; { ... };
    ($(<span class="attr">#[$(<span class="macro-nonterminal">$m</span>:meta),+]</span>)+ <span class="kw">pub enum </span><span class="macro-nonterminal">$e</span>:ident { $(<span class="macro-nonterminal">$v</span>:ident $(=<span class="macro-nonterminal">$val</span>:expr)<span class="kw-2">*</span>,)+ } ) =&gt; { ... };
    ($(<span class="attr">#[$(<span class="macro-nonterminal">$m</span>:meta),+]</span>)+ <span class="kw">pub enum </span><span class="macro-nonterminal">$e</span>:ident { $(<span class="macro-nonterminal">$v</span>:ident $(=<span class="macro-nonterminal">$val</span>:expr)<span class="kw-2">*</span>),+ } ) =&gt; { ... };
    ($(<span class="attr">#[$(<span class="macro-nonterminal">$m</span>:meta),+]</span>)+ <span class="kw">enum </span><span class="macro-nonterminal">$e</span>:ident { $(<span class="macro-nonterminal">$v</span>:ident $(=<span class="macro-nonterminal">$val</span>:expr)<span class="kw-2">*</span>,)+ } ) =&gt; { ... };
    ($(<span class="attr">#[$(<span class="macro-nonterminal">$m</span>:meta),+]</span>)+ <span class="kw">enum </span><span class="macro-nonterminal">$e</span>:ident { $(<span class="macro-nonterminal">$v</span>:ident $(=<span class="macro-nonterminal">$val</span>:expr)<span class="kw-2">*</span>),+ } ) =&gt; { ... };
    (<span class="kw">pub enum </span><span class="macro-nonterminal">$e</span>:ident { $(<span class="macro-nonterminal">$v</span>:ident $(=<span class="macro-nonterminal">$val</span>:expr)<span class="kw-2">*</span>,)+ } ) =&gt; { ... };
    (<span class="kw">pub enum </span><span class="macro-nonterminal">$e</span>:ident { $(<span class="macro-nonterminal">$v</span>:ident $(=<span class="macro-nonterminal">$val</span>:expr)<span class="kw-2">*</span>),+ } ) =&gt; { ... };
    (<span class="kw">enum </span><span class="macro-nonterminal">$e</span>:ident { $(<span class="macro-nonterminal">$v</span>:ident $(=<span class="macro-nonterminal">$val</span>:expr)<span class="kw-2">*</span>,)+ } ) =&gt; { ... };
    (<span class="kw">enum </span><span class="macro-nonterminal">$e</span>:ident { $(<span class="macro-nonterminal">$v</span>:ident $(=<span class="macro-nonterminal">$val</span>:expr)<span class="kw-2">*</span>),+ } ) =&gt; { ... };
}</pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Convenience macro to generate more complete enums with variants to be used as a type when
parsing arguments. This enum also provides a <code>variants()</code> function which can be used to
retrieve a <code>Vec&lt;&amp;'static str&gt;</code> of the variant names, as well as implementing <a href="https://doc.rust-lang.org/std/str/trait.FromStr.html"><code>FromStr</code></a> and
<a href="https://doc.rust-lang.org/std/fmt/trait.Display.html"><code>Display</code></a> automatically.</p>
<p><strong>NOTE:</strong> Case insensitivity is supported for ASCII characters only. It’s highly recommended to
use <a href="./struct.Arg.html#method.case_insensitive"><code>Arg::case_insensitive(true)</code></a> for args that will be used with these enums</p>
<p><strong>NOTE:</strong> This macro automatically implements <a href="https://doc.rust-lang.org/std/str/trait.FromStr.html"><code>std::str::FromStr</code></a> and <a href="https://doc.rust-lang.org/std/fmt/trait.Display.html"><code>std::fmt::Display</code></a></p>
<p><strong>NOTE:</strong> These enums support pub (or not) and uses of the <code>#[derive()]</code> traits</p>
<h2 id="examples"><a class="doc-anchor" href="#examples">§</a>Examples</h2>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="macro">arg_enum!</span>{
    <span class="attr">#[derive(PartialEq, Debug)]
    </span><span class="kw">pub enum </span>Foo {
        Bar,
        Baz,
        Qux
    }
}
<span class="comment">// Foo enum can now be used via Foo::Bar, or Foo::Baz, etc
// and implements std::str::FromStr to use with the value_t! macros
</span><span class="kw">fn </span>main() {
    <span class="kw">let </span>m = App::new(<span class="string">"app"</span>)
                .arg(Arg::from_usage(<span class="string">"&lt;foo&gt; 'the foo'"</span>)
                    .possible_values(<span class="kw-2">&amp;</span>Foo::variants())
                    .case_insensitive(<span class="bool-val">true</span>))
                .get_matches_from(<span class="macro">vec!</span>[
                    <span class="string">"app"</span>, <span class="string">"baz"
                </span>]);
    <span class="kw">let </span>f = <span class="macro">value_t!</span>(m, <span class="string">"foo"</span>, Foo).unwrap_or_else(|e| e.exit());

    <span class="macro">assert_eq!</span>(f, Foo::Baz);
}</code></pre></div>
</div></details></section></div></main></body></html>