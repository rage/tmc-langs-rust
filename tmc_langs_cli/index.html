<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="CLI client for TMC."><title>tmc_langs_cli - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-84e720fa.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="tmc_langs_cli" data-themes="" data-resource-suffix="" data-rustdoc-version="1.89.0 (29483883e 2025-08-04)" data-channel="1.89.0" data-search-js="search-92309212.js" data-settings-js="settings-5514c975.js" ><script src="../static.files/storage-4e99c027.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-fd3af306.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-32bb7600.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../tmc_langs_cli/index.html">tmc_<wbr>langs_<wbr>cli</a><span class="version">0.38.1</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section id="rustdoc-toc"><h3><a href="#modules">Crate Items</a></h3><ul class="block"><li><a href="#modules" title="Modules">Modules</a></li><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#enums" title="Enums">Enums</a></li><li><a href="#functions" title="Functions">Functions</a></li></ul></section><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Crate <span>tmc_langs_cli</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/tmc_langs_cli/lib.rs.html#1-1296">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>CLI client for TMC.</p>
</div></details><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="mod" href="app/index.html" title="mod tmc_langs_cli::app">app</a></dt><dd>Create clap app</dd><dt><a class="mod" href="error/index.html" title="mod tmc_langs_cli::error">error</a></dt><dd>Contains various error types.</dd><dt><a class="mod" href="output/index.html" title="mod tmc_langs_cli::output">output</a></dt><dd>Contains the type definition for the output format of the CLI.</dd></dl><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="struct" href="struct.CliError.html" title="struct tmc_langs_cli::CliError">CliError</a></dt></dl><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="enum" href="enum.ParsingResult.html" title="enum tmc_langs_cli::ParsingResult">Parsing<wbr>Result</a></dt></dl><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="fn" href="fn.map_parsing_result.html" title="fn tmc_langs_cli::map_parsing_result">map_<wbr>parsing_<wbr>result</a></dt><dt><a class="fn" href="fn.run.html" title="fn tmc_langs_cli::run">run</a></dt></dl></section></div></main></body></html>