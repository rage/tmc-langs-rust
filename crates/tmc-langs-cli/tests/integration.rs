use clap::Parser;
use std::path::Path;
use tempfile::{NamedTempFile, TempDir, tempdir};
use tmc_langs::{Compression, RunStatus, file_util};
use tmc_langs_cli::{
    app::Cli,
    output::{CliOutput, DataKind, OutputResult},
};
use walkdir::WalkDir;

fn cp_exercise(path: &Path) -> TempDir {
    let path_parent = path.parent().unwrap();
    let temp = tempdir().unwrap();
    for file in WalkDir::new(path) {
        let file = file.unwrap();
        let relative = file.path().strip_prefix(path_parent).unwrap();
        let target = temp.path().join(relative);
        if file.file_type().is_dir() {
            std::fs::create_dir_all(&target).unwrap();
        } else if file.file_type().is_file() {
            if let Some(parent) = target.parent() {
                std::fs::create_dir_all(parent).unwrap();
            }
            std::fs::copy(file.path(), &target).unwrap();
        }
    }
    temp
}

fn compress_naive(path: &impl AsRef<Path>, target: &impl AsRef<Path>, compression: Compression) {
    let cli = Cli::parse_from([
        "tmc-langs-cli",
        "--pretty",
        "compress-project",
        "--exercise-path",
        path_str(path),
        "--output-path",
        path_str(target),
        "--compression",
        &compression.to_string(),
        "--naive",
    ]);
    tmc_langs_cli::run(cli).unwrap();
}

fn extract_naive(path: &impl AsRef<Path>, target: &impl AsRef<Path>, compression: Compression) {
    let cli = Cli::parse_from([
        "tmc-langs-cli",
        "--pretty",
        "extract-project",
        "--archive-path",
        path_str(path),
        "--output-path",
        path_str(target),
        "--compression",
        &compression.to_string(),
        "--naive",
    ]);
    tmc_langs_cli::run(cli).unwrap();
}

fn path_str(path: &impl AsRef<Path>) -> &str {
    path.as_ref().to_str().unwrap()
}

fn sorted_list_of_files(path: &impl AsRef<Path>) -> Vec<String> {
    let mut files = vec![];
    for entry in WalkDir::new(path).min_depth(1) {
        let entry = entry.unwrap();
        let path = entry.path().strip_prefix(path.as_ref()).unwrap();
        let path = path_str(&path).replace('\\', "/");
        files.push(path);
    }
    files.sort();
    files
}

// wrapper for all sample exercise tests
fn test(f: impl Fn(&Path)) {
    let _ = env_logger::try_init();
    insta::with_settings!({
        filters => vec![
            // replace Windows-style path separators
            // the yaml serialization doubles the backslashes so
            // we need to filter \\\\
            (r"\\\\", "/"),
            // remove quotes, some keys and values get quoted on Windows due to backslash path separators
            ("\"", ""),

            // replace all /tmp/ (linux), /var/ (macos) and C:/[..]/Temp/ (win) paths which vary each test run
            // note that we already turned \s to /s
            (r"/tmp/\S*", "[PATH]"),
            (r"/var/\S*", "[PATH]"),
            (r"C:/\S*/Temp/\S*", "[PATH]"),
        ],
    }, {
        insta::glob!("../../../", "sample_exercises/*/*", |exercise| {
            let dir_name = exercise.file_name().unwrap();
            let exercise = cp_exercise(exercise);
            println!("testing {:?}", exercise.path().join(dir_name));
            f(&exercise.path().join(dir_name))
        })
    })
}

#[test]
fn checkstyle() {
    test(|exercise| {
        let out = NamedTempFile::new().unwrap();
        let cli = Cli::parse_from([
            "tmc-langs-cli",
            "--pretty",
            "checkstyle",
            "--exercise-path",
            path_str(&exercise),
            "--locale",
            "eng",
            "--output-path",
            path_str(&out),
        ]);
        let output = tmc_langs_cli::run(cli).unwrap();
        insta::assert_yaml_snapshot!(output);
    })
}

#[test]
fn clean() {
    test(|exercise| {
        let cli = Cli::parse_from([
            "tmc-langs-cli",
            "--pretty",
            "clean",
            "--exercise-path",
            path_str(&exercise),
        ]);
        let output = tmc_langs_cli::run(cli).unwrap();
        insta::assert_yaml_snapshot!(output);
        let files = sorted_list_of_files(&exercise);
        insta::assert_yaml_snapshot!(files);
    })
}

#[test]
#[cfg(not(target_os = "windows"))] // hash differs on windows
fn compress_project_tar() {
    test(|exercise| {
        let target = NamedTempFile::new().unwrap();
        let cli = Cli::parse_from([
            "tmc-langs-cli",
            "--pretty",
            "compress-project",
            "--exercise-path",
            path_str(&exercise),
            "--output-path",
            path_str(&target),
            "--compression",
            "tar",
            "--deterministic",
        ]);
        let output = tmc_langs_cli::run(cli).unwrap();
        insta::assert_yaml_snapshot!(output);

        let extracted = tempdir().unwrap();
        extract_naive(&target, &extracted, Compression::Tar);
        let files = sorted_list_of_files(&extracted);
        insta::assert_yaml_snapshot!(files);
    })
}

#[test]
#[cfg(not(target_os = "windows"))] // hash differs on windows
fn compress_project_zip() {
    test(|exercise| {
        let target = NamedTempFile::new().unwrap();
        let cli = Cli::parse_from([
            "tmc-langs-cli",
            "--pretty",
            "compress-project",
            "--exercise-path",
            path_str(&exercise),
            "--output-path",
            path_str(&target),
            // zip should be the default
            // "--compression",
            // "zip",
            "--deterministic",
        ]);
        let output = tmc_langs_cli::run(cli).unwrap();
        insta::assert_yaml_snapshot!(output);

        let extracted = tempdir().unwrap();
        extract_naive(&target, &extracted, Compression::Zip);
        let files = sorted_list_of_files(&extracted);
        insta::assert_yaml_snapshot!(files);
    })
}

#[test]
#[cfg(not(target_os = "windows"))] // hash differs on windows
fn compress_project_zstd() {
    test(|exercise| {
        let target = NamedTempFile::new().unwrap();
        let cli = Cli::parse_from([
            "tmc-langs-cli",
            "--pretty",
            "compress-project",
            "--exercise-path",
            path_str(&exercise),
            "--output-path",
            path_str(&target),
            "--compression",
            "zstd",
            "--deterministic",
        ]);
        let output = tmc_langs_cli::run(cli).unwrap();
        insta::assert_yaml_snapshot!(output);

        let extracted = tempdir().unwrap();
        extract_naive(&target, &extracted, Compression::TarZstd);
        let files = sorted_list_of_files(&extracted);
        insta::assert_yaml_snapshot!(files);
    })
}

#[test]
fn extract_project_tar() {
    test(|exercise| {
        let compressed = NamedTempFile::new().unwrap();
        compress_naive(&exercise, &compressed, Compression::Tar);
        let target = tempdir().unwrap();
        let cli = Cli::parse_from([
            "tmc-langs-cli",
            "--pretty",
            "extract-project",
            "--archive-path",
            path_str(&compressed),
            "--output-path",
            path_str(&target),
            "--compression",
            "tar",
        ]);
        let output = tmc_langs_cli::run(cli).unwrap();
        insta::assert_yaml_snapshot!(output);
        let files = sorted_list_of_files(&target);
        insta::assert_yaml_snapshot!(files);
    })
}

#[test]
fn extract_project_zip() {
    test(|exercise| {
        let compressed = NamedTempFile::new().unwrap();
        compress_naive(&exercise, &compressed, Compression::Zip);
        let target = tempdir().unwrap();
        let cli = Cli::parse_from([
            "tmc-langs-cli",
            "--pretty",
            "extract-project",
            "--archive-path",
            path_str(&compressed),
            "--output-path",
            path_str(&target),
            // zip should be the default
            // "--compression",
            // "zip",
        ]);
        let output = tmc_langs_cli::run(cli).unwrap();
        insta::assert_yaml_snapshot!(output);
        let files = sorted_list_of_files(&target);
        insta::assert_yaml_snapshot!(files);
    })
}

#[test]
fn extract_project_zstd() {
    test(|exercise| {
        let compressed = NamedTempFile::new().unwrap();
        compress_naive(&exercise, &compressed, Compression::TarZstd);
        let target = tempdir().unwrap();
        let cli = Cli::parse_from([
            "tmc-langs-cli",
            "--pretty",
            "extract-project",
            "--archive-path",
            path_str(&compressed),
            "--output-path",
            path_str(&target),
            "--compression",
            "zstd",
        ]);
        let output = tmc_langs_cli::run(cli).unwrap();
        insta::assert_yaml_snapshot!(output);
        let files = sorted_list_of_files(&target);
        insta::assert_yaml_snapshot!(files);
    })
}

#[test]
fn prepare_submission_tar() {
    test(|exercise| {
        let compressed = NamedTempFile::new().unwrap();
        compress_naive(&exercise, &compressed, Compression::Tar);
        let target = NamedTempFile::new().unwrap();
        let cli = Cli::parse_from([
            "tmc-langs-cli",
            "--pretty",
            "prepare-submission",
            "--clone-path",
            path_str(&exercise),
            "--output-path",
            path_str(&target),
            // tar should be the default
            // "--output-format",
            // "tar",
            "--submission-path",
            path_str(&compressed),
            "--submission-compression",
            "tar",
            "--no-archive-prefix",
        ]);
        let output = tmc_langs_cli::run(cli).unwrap();
        insta::assert_yaml_snapshot!(output);

        let extract_target = tempdir().unwrap();
        extract_naive(&target, &extract_target, Compression::Tar);
        let files = sorted_list_of_files(&extract_target);
        insta::assert_yaml_snapshot!(files);
    })
}

#[test]
fn prepare_submission_zip() {
    test(|exercise| {
        let compressed = NamedTempFile::new().unwrap();
        compress_naive(&exercise, &compressed, Compression::Zip);
        let target = NamedTempFile::new().unwrap();
        let cli = Cli::parse_from([
            "tmc-langs-cli",
            "--pretty",
            "prepare-submission",
            "--clone-path",
            path_str(&exercise),
            "--output-path",
            path_str(&target),
            "--output-format",
            "zip",
            "--submission-path",
            path_str(&compressed),
            // "--submission-compression",
            // "zip",
            "--no-archive-prefix",
        ]);
        let output = tmc_langs_cli::run(cli).unwrap();
        insta::assert_yaml_snapshot!(output);

        let extract_target = tempdir().unwrap();
        extract_naive(&target, &extract_target, Compression::Zip);
        let files = sorted_list_of_files(&extract_target);
        insta::assert_yaml_snapshot!(files);
    })
}

#[test]
fn prepare_submission_zstd() {
    test(|exercise| {
        let compressed = NamedTempFile::new().unwrap();
        compress_naive(&exercise, &compressed, Compression::TarZstd);
        let target = NamedTempFile::new().unwrap();
        let cli = Cli::parse_from([
            "tmc-langs-cli",
            "--pretty",
            "prepare-submission",
            "--clone-path",
            path_str(&exercise),
            "--output-path",
            path_str(&target),
            "--output-format",
            "zstd",
            "--submission-path",
            path_str(&compressed),
            "--submission-compression",
            "zstd",
            "--no-archive-prefix",
        ]);
        let output = tmc_langs_cli::run(cli).unwrap();
        insta::assert_yaml_snapshot!(output);

        let extract_target = tempdir().unwrap();
        extract_naive(&target, &extract_target, Compression::TarZstd);
        let files = sorted_list_of_files(&extract_target);
        insta::assert_yaml_snapshot!(files);
    })
}

#[test]
// dirs returns wrong user's cache dir on github's windows runner which causes issues with maven plugin
// todo: use python exercise as base instead
#[cfg(not(target_os = "windows"))]
fn prepare_submission_fails_tests() {
    let _ = env_logger::try_init();

    println!("set up user submission");
    let submission_super = tempfile::tempdir().unwrap();
    let exercise_clone = "../../sample_exercises/java/maven-exercise";
    let exercise_model = "../../sample_exercises/java/maven-exercise-model";
    file_util::copy(exercise_clone, &submission_super.path()).unwrap();
    let submission_dir = submission_super.path().join("maven-exercise");
    let cli = Cli::parse_from([
        "tmc-langs-cli",
        "--pretty",
        "run-tests",
        "--exercise-path",
        path_str(&submission_dir),
    ]);
    let output = tmc_langs_cli::run(cli).unwrap();

    println!("ensure unmodified exercise fails");
    match output {
        CliOutput::OutputData(data) => match data.data {
            Some(DataKind::TestResult(result)) => {
                print!("{:#?}", result.logs);
                assert_eq!(result.status, RunStatus::TestsFailed);
            }
            _ => panic!(),
        },
        _ => panic!(),
    }

    println!("compress");
    let compressed_project = tempfile::NamedTempFile::new().unwrap();
    let cli = Cli::parse_from([
        "tmc-langs-cli",
        "--pretty",
        "compress-project",
        "--exercise-path",
        path_str(&submission_dir),
        "--output-path",
        path_str(&compressed_project.path()),
    ]);
    let output = tmc_langs_cli::run(cli).unwrap();
    match output {
        CliOutput::OutputData(data) => {
            assert_eq!(data.result, OutputResult::ExecutedCommand);
        }
        _ => panic!(),
    }

    println!("prepare submission");
    let prepared_tar = tempfile::NamedTempFile::new().unwrap();
    let cli = Cli::parse_from([
        "tmc-langs-cli",
        "--pretty",
        "prepare-submission",
        "--clone-path",
        exercise_model,
        "--submission-path",
        path_str(&compressed_project.path()),
        "--output-path",
        path_str(&prepared_tar.path()),
        "--no-archive-prefix",
    ]);
    let output = tmc_langs_cli::run(cli).unwrap();
    match output {
        CliOutput::OutputData(data) => {
            assert_eq!(data.result, OutputResult::ExecutedCommand);
        }
        _ => panic!(),
    }

    println!("extract tar at {}", prepared_tar.path().display());
    let submission_prepared = tempfile::TempDir::new().unwrap();
    let cli = Cli::parse_from([
        "tmc-langs-cli",
        "--pretty",
        "extract-project",
        "--naive",
        "--archive-path",
        path_str(&prepared_tar.path()),
        "--compression",
        "tar",
        "--output-path",
        path_str(&submission_prepared.path()),
    ]);
    let output = tmc_langs_cli::run(cli).unwrap();
    match output {
        CliOutput::OutputData(data) => {
            assert_eq!(data.result, OutputResult::ExecutedCommand);
        }
        _ => panic!(),
    }

    println!("test submission");
    let cli = Cli::parse_from([
        "tmc-langs-cli",
        "--pretty",
        "run-tests",
        "--exercise-path",
        path_str(&submission_prepared.path()),
    ]);
    let output = tmc_langs_cli::run(cli).unwrap();

    println!("ensure exercise still fails after prep");
    match output {
        CliOutput::OutputData(data) => match data.data {
            Some(DataKind::TestResult(result)) => {
                println!("{:#?}", result.logs);
                println!("{:#?}", result.test_results);
                assert_eq!(result.status, RunStatus::TestsFailed);
            }
            _ => panic!(),
        },
        _ => panic!(),
    }
}

#[test]
// dirs returns wrong user's cache dir on github's windows runner which causes issues with maven plugin
// todo: use python exercise as base instead
#[cfg(not(target_os = "windows"))]
fn prepare_submission_passes_tests() {
    let _ = env_logger::try_init();

    println!("set up user submission");
    let temp_maven = tempfile::tempdir().unwrap();
    let maven_exercise = "../../sample_exercises/java/maven-exercise-model";
    file_util::copy(maven_exercise, &temp_maven.path()).unwrap();
    let user_path = temp_maven.path().join("maven-exercise-model");
    let cli = Cli::parse_from([
        "tmc-langs-cli",
        "--pretty",
        "run-tests",
        "--exercise-path",
        path_str(&user_path),
    ]);
    let output = tmc_langs_cli::run(cli).unwrap();

    println!("ensure unmodified exercise passes");
    match output {
        CliOutput::OutputData(data) => match data.data {
            Some(DataKind::TestResult(result)) => {
                print!("{:#?}", result.logs);
                assert_eq!(result.status, RunStatus::Passed);
            }
            _ => panic!(),
        },
        _ => panic!(),
    }

    println!("compress");
    let temp_compressed = tempfile::NamedTempFile::new().unwrap();
    let cli = Cli::parse_from([
        "tmc-langs-cli",
        "--pretty",
        "compress-project",
        "--exercise-path",
        path_str(&user_path),
        "--output-path",
        path_str(&temp_compressed.path()),
    ]);
    let output = tmc_langs_cli::run(cli).unwrap();
    match output {
        CliOutput::OutputData(data) => {
            assert_eq!(data.result, OutputResult::ExecutedCommand);
        }
        _ => panic!(),
    }

    println!("prepare submission");
    let temp_tar = tempfile::NamedTempFile::new().unwrap();
    let cli = Cli::parse_from([
        "tmc-langs-cli",
        "--pretty",
        "prepare-submission",
        "--clone-path",
        maven_exercise,
        "--submission-path",
        path_str(&temp_compressed.path()),
        "--output-path",
        path_str(&temp_tar.path()),
        "--no-archive-prefix",
    ]);
    let output = tmc_langs_cli::run(cli).unwrap();
    match output {
        CliOutput::OutputData(data) => {
            assert_eq!(data.result, OutputResult::ExecutedCommand);
        }
        _ => panic!(),
    }

    println!("extract tar at {}", temp_tar.path().display());
    let submission_finished = tempfile::TempDir::new().unwrap();
    let cli = Cli::parse_from([
        "tmc-langs-cli",
        "--pretty",
        "extract-project",
        "--naive",
        "--archive-path",
        path_str(&temp_tar.path()),
        "--compression",
        "tar",
        "--output-path",
        path_str(&submission_finished.path()),
    ]);
    let output = tmc_langs_cli::run(cli).unwrap();
    match output {
        CliOutput::OutputData(data) => {
            assert_eq!(data.result, OutputResult::ExecutedCommand);
        }
        _ => panic!(),
    }

    println!("test submission");
    let cli = Cli::parse_from([
        "tmc-langs-cli",
        "--pretty",
        "run-tests",
        "--exercise-path",
        path_str(&submission_finished.path()),
    ]);
    let output = tmc_langs_cli::run(cli).unwrap();

    println!("ensure exercise passes after prep");
    match output {
        CliOutput::OutputData(data) => match data.data {
            Some(DataKind::TestResult(result)) => {
                println!("{:#?}", result.logs);
                println!("{:#?}", result.test_results);
                assert_eq!(result.status, RunStatus::Passed);
            }
            _ => panic!(),
        },
        _ => panic!(),
    }
}
