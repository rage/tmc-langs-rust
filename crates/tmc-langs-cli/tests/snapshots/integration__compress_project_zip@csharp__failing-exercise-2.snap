---
source: crates/tmc-langs-cli/tests/integration.rs
expression: files
input_file: sample_exercises/csharp/failing-exercise
---
- failing-exercise
- failing-exercise/src
- failing-exercise/src/FailingSample
- failing-exercise/src/FailingSample/Program.cs
